Client/Frontend/Sync/BraveCore/Bookmarkv2.swift

```diff
@@ -0,0 +1,316 @@
+// Copyright 2020 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import Data
+import BraveRewards
+import CoreData
+
+// A Lightweight wrapper around BraveCore bookmarks
+// with the same layout/interface as `Bookmark (from CoreData)`
+class Bookmarkv2 {
+    private let bookmarkNode: BookmarkNode?
+    private static let bookmarksAPI = BraveBookmarksAPI()
+    
+    init(_ bookmarkNode: BookmarkNode) {
+        self.bookmarkNode = bookmarkNode
+    }
+    
+    public var isFolder: Bool {
+        return bookmarkNode?.isFolder == true
+    }
+    
+    public var title: String? {
+        return bookmarkNode?.titleUrlNodeTitle
+    }
+    
+    public var customTitle: String? {
+        return bookmarkNode?.titleUrlNodeTitle
+    }
+    
+    public var displayTitle: String? {
+        return self.customTitle
+    }
+    
+    public var url: String? {
+        return bookmarkNode?.titleUrlNodeUrl?.absoluteString
+    }
+    
+    public var domain: Domain? {
+        if let url = bookmarkNode?.url {
+            return Domain.getOrCreate(forUrl: url, persistent: true)
+        }
+        return nil
+    }
+    
+    public var created: Date? {
+        get {
+            return bookmarkNode?.dateAdded
+        }
+        
+        set {
+            bookmarkNode?.dateAdded = newValue ?? Date()
+        }
+    }
+    
+    public var parent: Bookmarkv2? {
+        get {
+            if let parent = bookmarkNode?.parent {
+                return Bookmarkv2(parent)
+            }
+            return nil
+        }
+    }
+    
+    public var children: [Bookmarkv2]? {
+        return bookmarkNode?.children.map({ Bookmarkv2($0) })
+    }
+    
+    public var canBeDeleted: Bool {
+        return bookmarkNode?.isPermanentNode == false
+    }
+    
+    public var objectID: Int {
+        return Int(bookmarkNode?.nodeId ?? 0)
+    }
+    
+    public var order: Int16 {
+        let children = bookmarkNode?.parent?.children
+        return Int16(children?.firstIndex(where: { $0.guid == self.bookmarkNode?.guid }) ?? -1)
```

Comment:
It shouldn't be possible for this to happen ever. A bookmark on the Chromium side always has a parent.. There is only one node possible that can have no parent and that is `Root`.

It goes:

```
Root {
   MobileBookmarks,
   DesktopBookmarks,
   OtherBookmarks
}
```

Root is never exposed in this API at all so this should never be the case for iOS. I only added the `-1` there because the field is not optional.